utils/thapi_gen_source_callbacks.rb

```diff
@@ -0,0 +1,205 @@
+
+  void btx_push_usr_messages(void *btx_handle, void *usr_data, btx_source_status_t *status) {
+      <%- if not data.empty? and not data.first[:hostname].nil? -%>
+      btx_downstream_set_environment_hostname(btx_handle, "<%= data.first[:hostname] %>");
+      <%- end -%>
+
+      <%- data.each do | entry | -%>
+      <%- entry.fetch(:times,1).times do -%>
+      btx_push_message_<%= entry[:name] %>(btx_handle<%= ', ' if not entry[:field_values].empty? %><%= 
+       entry[:field_casts].zip(entry[:field_values].map{ |v| v.to_s }).map{ |cv| '(' + cv[0] + ')' + cv[1] }.join(', ')
+      %>);
+      <%- end -%>
+      <%- end -%>
+
+      *status = BTX_SOURCE_END;
+  }
+
+  void btx_register_usr_callbacks(void *btx_handle) {
+    btx_register_callbacks_push_usr_messages(btx_handle, &btx_push_usr_messages);
+  }
+TEXT
+
+BASE_DATE = Time.utc(2024)
+
+def get_values(name, value)
+  if not name.end_with?("_vals") or not value.strip.start_with?("[")
+    # simple value, do not modify
+    return [value]
+  end
+
+  # value will look like '[ ]' or '[ value ]' or '[ val1, val2, ... ]'
+  count = 0
+  value = value.strip[1...-1].strip
+  if value.empty?
+    count = 0
+    # use static const empty array defined in callback template
+    value = "EMPTY_ARRAY";
+  else
+    values = value.split(/, */)
+    count = 1
+    # TODO: handle multiple values
+    value = values[0]
+  end
+  [count, value]
+  # hack: always return empty array
+  [0, "EMPTY_ARRAY"]
+end
+
+def parse_field_values(fields_string, exclude_fields)
+  # expects commas separated list of name: value pairs
+  values = []
+  fields_string.split(/, */).each do |nv_pair|
+    name, value = nv_pair.split(/: /)
+    name.strip!
+    if not exclude_fields.member?(name)
+      values += get_values(name, value)
+    end
+  end
+  values
+end
+
+def get_fields(model, event_name, stream_class=0)
+  fields = {}
+  stream_classes = model[:stream_classes]
+  sc = stream_classes[0]
+  sc[:event_common_context_field_class][:members].each do |field_info|
+    fields[field_info[:name]] = field_info[:field_class][:cast_type]
+  end
+  sc[:event_classes].each do |eclass|
+    if eclass[:name] != event_name
+      next
+    end
+    eclass[:payload_field_class][:members].each do |field_info|
+      fields[field_info[:name]] = field_info[:field_class][:cast_type]
+    end
+    break
+  end
+  fields
+end
+
+def combine_date_time(date, time)
+  Time.utc(date.year, date.month, date.mday,
+           time.hour, time.min, time.sec, time.subsec)
+end
+
+# Example input line from babeltrace_thapi:
+# 12:59:07.177888481 - gene-lxc - vpid: 415, vtid: 415 - lttng_ust_cuda:cuDevicePrimaryCtxRetain_exit: { cuResult: CUDA_SUCCESS, pctx_val: 0x000055c6244d84f0 }
+def parse_event(model, line, exclude_fields)
+  h = { hostname: nil,
+        name: nil,
+        field_values: [],
+        field_casts: [] }
+  # First 18 characters are the timestmap
+  t = combine_date_time(BASE_DATE, Time.parse(line[0..17]))
```

Comment:
Maybe `.shift(17)` and then `.shift(3)` will make revirewer does less math :)

Reply:
Doesn't look like shift is available on String, I guess could do something `line = line[18...]`, or call it tail.

Reply:
maybe simplest is to just split everything on ' - ', there is no performance issue here.

Reply:
oh yeah need to do `str.chars.shift`, agree split maybe eathier :)